writefile.go

```diff
@@ -20,18 +20,18 @@ import "os"
 
 // WriteFile mirrors ioutil.WriteFile, replacing an existing file with the same
 // name atomically.
-func WriteFile(filename string, data []byte, perm os.FileMode) error {
-	t, err := TempFile("", filename)
+func WriteFile(filename string, data []byte, perm os.FileMode, opts ...Option) error {
+	opts = append([]Option{
+		WithPermissions(perm),
+		WithExistingPermissions(),
```

Comment:
Are we sure that `WithExistingPermissions()` is good to add as a default option here?

`ioutil.WriteFile` doesn’t do this, right? So it’d be a divergence.

It’s also an extra syscall, which can be relevant in terms of overhead.

Also, how can users disable this code path? They’d need to use the more verbose `NewPendingFile` entrypoint, right?

Reply:
> `ioutil.WriteFile` doesn’t do this, right?

Yes and no. `ioutil.WriteFile` and its newer location `os.WriteFile` call `os.CreateFile` with `O_WRONLY|O_CREATE|O_TRUNC` and permissions (https://cs.opensource.google/go/go/+/refs/tags/go1.17.1:src/os/file.go;l=713-723;drc=refs%2Ftags%2Fgo1.17.1). The operating system will reuse existing permissions if the file already exists. The given permissions are only used for creating files—after applying the umask.

`WithExistingPermissions` is necessary to preserve that behaviour in `renameio`: the temporary file is created/configured with the permissions from the target. As with `ioutil.WriteFile` there's no way around this currently.

The tests in `writefile_test.go:TestWriteFileEquivalence` verify that both functions, `ioutil.WriteFile` and `renameio.WriteFile`, behave the same way (without extra options to `renameio.WriteFile`).

Reply:
Thank you for explaining!